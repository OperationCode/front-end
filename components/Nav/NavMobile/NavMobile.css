.NavMobile {
  position: fixed;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background-color: var(--white);
  height: var(--navBarHeight);
  width: 100%;

  --gutterWidth: 2rem;
}

.logoLink {
  margin-left: var(--gutterWidth);
  max-width: 50%;
  max-height: calc(var(--navBarHeight) - 1rem);
}

.logo {
  max-width: 100%;
  max-height: calc(var(--navBarHeight) - 1rem);
  vertical-align: middle;
}

.hamburgerButton {
  margin-right: var(--gutterWidth);
  border: none;
  background: none;
}

.hamburgerButton:hover {
  cursor: pointer;
}

.hamburgerIcon {
  width: 35px;
}

.dropdown {
  position: absolute;
  left: 0;
  top: 0;
  background: rgba(37, 46, 62, 0.97);
  width: 97.5%;
  height: 100vh;
  z-index: 10;
}

.ul {
  padding: 1rem;
  margin: 0;
  list-style: none;
  font-size: 1.5rem;
  overflow-y: auto;
}

.li {
  padding: 1rem;
}

.li:first-child {
  padding-top: 2rem;
}

.li:last-child {
  padding-bottom: 3rem;
}

.a {
  color: var(--white);
  fill: var(--white);
  transition: color 0.2s linear, fill 0.2s linear;
}

.a:hover {
  color: var(--primary);
  fill: var(--primary);
  transition: color 0.2s linear, fill 0.2s linear;
}
