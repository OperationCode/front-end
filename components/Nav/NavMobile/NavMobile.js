import React, { Component } from 'react';
import Link from 'next/link';
import { donateLink, s3 } from 'common/constants/urls';
import { navItems } from 'common/constants/navigation';
import flattenDepth from 'lodash/flattenDepth';
import HamburgerIcon from 'static/images/icons/hamburger.svg';
import CloseButton from 'components/_common_/CloseButton/CloseButton';
import styles from './NavMobile.css';

export default class NavMobile extends Component {
  state = {
    isDropDownVisible: false,
  };

  openDropDown = () => this.setState({ isDropDownVisible: true });

  closeDropDown = () => this.setState({ isDropDownVisible: false });

  render() {
    const { isDropDownVisible } = this.state;

    const links = flattenDepth(navItems.map(navItem => [navItem, navItem.sublinks]), 2);

    return (
      <header className={styles.NavMobile}>
        <Link href="/">
          <a className={styles.logoLink}>
            <img
              src={`${s3}branding/logos/small-blue-logo.png`}
              alt="Operation Code Logo"
              className={styles.logo}
            />
          </a>
        </Link>

        <button
          className={styles.hamburgerButton}
          type="button"
          name="dropdown"
          onClick={this.openDropDown}
        >
          <HamburgerIcon className={styles.hamburgerIcon} />
        </button>

        {isDropDownVisible && (
          <nav className={styles.dropdown}>
            <CloseButton onClick={this.closeDropDown} theme="white" />

            <ul className={styles.ul}>
              <li className={styles.li}>
                <Link href="/">
                  <a className={styles.a}>Home</a>
                </Link>
              </li>
              {links.map(navlink => (
                <li className={styles.li} key={navlink.name}>
                  <Link href={navlink.href}>
                    <a className={styles.a}>{navlink.name}</a>
                  </Link>
                </li>
              ))}
              <li className={styles.li}>
                <Link href={donateLink}>
                  <a className={styles.a}>Donate</a>
                </Link>
              </li>
            </ul>
          </nav>
        )}
      </header>
    );
  }
}
