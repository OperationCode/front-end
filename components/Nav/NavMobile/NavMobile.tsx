import Link from 'next/link';
import classNames from 'classnames';
import { s3 } from '@/common/constants/urls';
import HamburgerIcon from '@/public/static/images/icons/hamburger.svg';
import { CloseButton } from '@/components/Buttons/CloseButton/CloseButton';
import { ScreenReaderOnly } from '@/components/ScreenReaderOnly/ScreenReaderOnly';
import Image from 'next/legacy/image';
import styles from './NavMobile.module.css';

interface SublinkType {
  /**
   * String used as the link label.
   */
  name: string;
  /**
   * String used for the URL.
   */
  href: string;
}

interface NavItemType {
  /**
   * String used as the link label.
   */
  name: string;

  /**
   * String used for the URL.
   */
  href: string;
  /**
   * Adds nested sublinks.
   */
  sublinks?: SublinkType[];
}

export interface NavMobilePropsType {
  /**
   * Sets if the mobile navigation is open or closed.
   */
  isOpen: boolean;
  /**
   * Function called when open button is clicked.
   */
  openMenu: () => void;
  /**
   * Function called when close button is clicked.
   */
  closeMenu: () => void;
  /**
   * List of navigations items.
   */
  navItems: NavItemType[];
}

export function NavMobile({ isOpen, openMenu, closeMenu, navItems }: NavMobilePropsType) {
  return (
    <header className={styles.NavMobile} data-testid="Mobile Nav Container">
      <Link href="/" className={classNames(styles.button, styles.logoButton, 'inline-flex')}>
        <div className={styles.logo}>
          <Image
            src={`${s3}branding/logos/small-blue-logo.png`}
            alt="Operation Code Logo"
            width={224}
            height={42}
          />
        </div>
      </Link>

      <button
        className={classNames(styles.button, styles.hamburger)}
        onClick={openMenu}
        type="button"
        name="dropdown"
        data-testid="Hamburger Button"
      >
        <ScreenReaderOnly>Open Menu</ScreenReaderOnly>
        <HamburgerIcon className={styles.hamburgerIcon} />
      </button>

      {isOpen && (
        <nav data-testid="Mobile Nav">
          <CloseButton onClick={closeMenu} theme="white" />

          <ul className={styles.ul}>
            <li className={styles.li} key="Home">
              <Link href="/" className={styles.link}>
                Home
              </Link>
            </li>

            {navItems.map(navlink => (
              <li className={styles.li} key={navlink.name}>
                <Link href={navlink.href} className={styles.link}>
                  {navlink.name}
                </Link>
              </li>
            ))}
          </ul>
        </nav>
      )}
    </header>
  );
}
