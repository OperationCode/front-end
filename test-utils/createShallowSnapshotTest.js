import ShallowRenderer from 'react-test-renderer/shallow';

/**
 * Used to create a snapshot test off of surface leaves of the tree.
 * Useful for regression tests on containers, components utilizing something that requires large
 * mocking, or very large components.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const shallowRenderer = new ShallowRenderer();
  const tree = shallowRenderer.render(Component);

  expect(tree).toMatchSnapshot();
};
