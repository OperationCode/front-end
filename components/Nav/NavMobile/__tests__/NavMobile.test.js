import React from 'react';
import { shallow } from 'enzyme';
import createShallowSnapshotTest from 'test-utils/createShallowSnapshotTest';

import NavMobile from '../NavMobile';

describe('NavMobile', () => {
  it('should render', () => createShallowSnapshotTest(<NavMobile />));

  it('should not immediately render nav items list', () => {
    const wrapper = shallow(<NavMobile />);

    expect(wrapper.find('ul')).not.toExist();
  });

  it('should render nav items list after hamburger button is clicked', () => {
    const wrapper = shallow(<NavMobile />);

    wrapper.find('.hamburgerButton').simulate('click');

    expect(wrapper.find('ul')).toExist();
  });

  it('should hide nav items list after close button is pressed', () => {
    const wrapper = shallow(<NavMobile />);

    wrapper.setState({ isDropDownVisible: true });
    wrapper.update();

    expect(wrapper.find('ul')).toExist();

    wrapper.find('CloseButton').simulate('click');
    expect(wrapper.find('ul')).not.toExist();
  });
});
