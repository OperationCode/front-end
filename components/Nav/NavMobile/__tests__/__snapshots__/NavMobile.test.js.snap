// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMobile should render 1`] = `
<header
  className="NavMobile"
>
  <Link
    href="/"
  >
    <a
      className="logoLink"
    >
      <img
        alt="Operation Code Logo"
        className="logo"
        src="https://s3.amazonaws.com/operationcode-assets/branding/logos/small-blue-logo.png"
      />
    </a>
  </Link>
  <button
    className="hamburgerButton"
    name="dropdown"
    onClick={[Function]}
    type="button"
  >
    <Svg
      className="hamburgerIcon"
    />
  </button>
</header>
`;
